File: 157. Read N Characters Given Read4.py
"""
The read4 API is already defined for you.

    @param buf, a list of characters
    @return an integer
    def read4(buf):

# Below is an example of how the read4 API can be called.
file = File("abcdefghijk") # File is "abcdefghijk", initially file pointer (fp) points to 'a'
buf = [' '] * 4 # Create buffer with enough space to store characters
read4(buf) # read4 returns 4. Now buf = ['a','b','c','d'], fp points to 'e'
read4(buf) # read4 returns 4. Now buf = ['e','f','g','h'], fp points to 'i'
read4(buf) # read4 returns 3. Now buf = ['i','j','k',...], fp points to end of file
"""
class Solution:
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Number of characters to read (int)
        :rtype: The number of actual characters read (int)
        """
        leng = 0
        buf4 = [' ', ' ', ' ', ' ']
        while n > 4:
            r4 = read4(buf4)
            if r4 < 4:
                buf[leng:leng+r4] = buf4[:r4]
                leng += r4
                return leng
            else:
                buf[leng:leng+4] = buf4[:]
                leng += 4
                n -= 4
        r4 = read4(buf4)
        if r4 < n:
            buf[leng:leng+r4] = buf4[:r4]
            leng += r4
            return leng
        else:
            buf[leng:leng+n] = buf4[:n]
            leng += n
            return leng

File: 157. Read N Characters Given Read4.java
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char[] buf);
 */
public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    public int read(char[] buf, int n) {
        int curr = 0;
        char[] b = new char[4];
        while (curr < n) {
            int ret = read4(b);
            if (ret < 4 && curr+ret<=n) {
                int count = 0;
                while (count < ret) {
                    buf[curr++] = b[count++];
                }
                return curr;
            } else if (curr+ret>n) {
                int count = 0;
                while (curr < n) {
                    buf[curr++] = b[count++];
                }
                return n;
            } else {
                int count = 0;
                while (count < 4) {
                    buf[curr++] = b[count++];
                }
            }
        }
        return n;
    }
}